pkg/engine/internal/worker/worker.go

```diff
@@ -171,7 +167,9 @@ func (w *Worker) Service() services.Service {
 
 // run starts the worker, running until the provided context is canceled.
 func (w *Worker) run(ctx context.Context) error {
-	g, ctx := errgroup.WithContext(ctx)
+	threadsCtx, threadsCancel := context.WithCancel(context.Background())
```

Comment:
we create multiple new contexts now, all separated from the the provided ctx argument. is there a risk in having these all be independent? it does make for some more complicated shutdown code as well, but maybe this separation is necessary/required?